Unit tests

batiati opened the pull request:
- Finishes all pending unit tests

We achieved 96% of test coverage, only a few paths cannot be tested without tampering with the implementation itself.
![image](https://user-images.githubusercontent.com/5728002/191070513-c0a472f7-fb9c-4ebc-ab12-07745070a6ce.png)

- Adds Javadocs

The main purpose of adding Javadocs now is to document parameters and exceptions.
I will revisit them once the official docs are released.

batiati committed:
Account

- Unit tests for all acessors
- Javadoc describing all acessors, parameters, return values and exceptions

batiati committed:
More tests

- Transfer acessors tests
- Transfer Javadocs
- More coverage for existing tests
- Removed wrong "throws" block

batiati committed:
Minor refactory

- Remove wrong "throws" block
- Rephrase comments

batiati committed:
More tests

- Request unit tests
- Removed wrong "throws" block
- More javadoc

batiati committed:
More unit tests

- Constants are now declared on interfaces, instead of "final classes"
It is more idiomatic

- More asserts and parameter validations

- More javadocs

batiati committed:
More unit tests

- JNILoader tests
- More Javadocs

batiati committed:
Minor adjustments

- Add a message to all NullPointerException
- Remove random values from tests

batiati committed:
More tests

- Adds missing test for TOO_MUCH_DATA result
- Adds test for a null element on the replicaAddresses array

Javados

- Updates javadoc on NullPointerException regarding null arrays and null elements.

Fixes

- Fixes the assert order for "buffer == null", since it's expected to receive a null bullfer when "status != OK"

Minor rework

- Uses "import statis" instead of "Assert.assertXYZ" on IntegrationTests.java, following the other tests.
- Removes the "exception swallowing" on "IntegrationTests.Server.cleanUp"

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Account.java:
```diff
@@ -50,99 +56,208 @@ public Account() {
+     *        unique 128-bit integer.
+     * @throws NullPointerException if {@code id} is null.
+     */
+    public void setId(UUID id) {
```
It is not required that the id is a UUID. So I'm not sure we should do this.

Same for get/setUserData. UUID is *one* valid kind of id and userData but it is not the only one. It is just an example.

jorangreef replied:
100% — it should be a 128-bit integer only as far as possible. The most primitive type.

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Account.java:
```diff
@@ -50,99 +56,208 @@ public Account() {
     public void setFlags(int flags) {
         if (flags < 0 || flags > Character.MAX_VALUE)
-            throw new IllegalArgumentException("Flags must be a unsigned 16 bits value");
+            throw new IllegalArgumentException("Flags must be a 16-bits unsigned integer");
```
I'd say `16-bit unsigned integer` not `16-bits unsigned integer`. I can't explain it but the former seems more natural

Everywhere you say X-bits unsigned integer it should probably be X-bit unsigned integer

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Account.java:
```diff
@@ -50,99 +56,208 @@ public Account() {
+     *
+     * @return A 64-bits integer.
+     */
     public long getDebitsPending() {
```
I wonder if it's a big deal that Java doesn't have signed integers.

In the Node client we use BigInteger instead of the default number type.

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -26,43 +20,103 @@ public final class Client implements AutoCloseable {
+     * @param replicaAddresses
+     * @param maxConcurrency
+     *
+     * @throws InitializationException If TigerBeetle an error occurred initializing this client.
```
"If TigerBeetle an error" doesn't make sense. Is "TigerBeetle" supposed to be in this sentence?

batiati replied:
Copy and paste crap, sorry ;-)

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -26,43 +20,103 @@ public final class Client implements AutoCloseable {
+     * @throws InitializationException If TigerBeetle an error occurred initializing this client.
+     *         See {@link InitializationException.Status} for more details.
+     *
+     * @throws IllegalArgumentException If {@code clusterID} is negative.
```
You've been inconsistent in this PR about capitalize `if/If` after `@throws`. I'm not sure which one it should be but they should probably all be the same?

batiati replied:
They should be capitalized only in top-level comments, never after tags like `@param`, `@return`, and `@throws`.

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -26,43 +20,103 @@ public final class Client implements AutoCloseable {
+     *
+     * @throws IllegalArgumentException If {@code clusterID} is negative.
+     * @throws IllegalArgumentException If {@code replicaAddresses} is empty or presented in
+     *         incorrect format.
```
You didn't describe the _correct_ format in this docstring if that is something you intended to do

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -26,43 +20,103 @@ public final class Client implements AutoCloseable {
+     * @param clusterID
+     * @param replicaAddresses
+     *
+     * @throws InitializationException If TigerBeetle an error occurred initializing this client.
```
Same comment about the sentence. "If TigerBeetle an error" doesn't make sense

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -26,43 +20,103 @@ public final class Client implements AutoCloseable {
+     * Submits a new account to be created.
+     *
+     * @param account A single {@link com.tigerbeetle.Account} instance to be created.
+     * @return A {@link com.tigerbeetle.CreateAccountResult}
```
Missing a period?

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -75,30 +129,86 @@ public CreateAccountResult createAccount(Account account) throws RequestExceptio
     }
 
+    /**
+     * Looks up for a single account.
```
"Looks up a single account." No "for"

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -108,27 +218,79 @@ public Account lookupAccount(UUID uuid) throws RequestException {
     }
 
+    /**
+     * Looks up for a batch of accounts.
```
No "for"

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -108,27 +218,79 @@ public Account lookupAccount(UUID uuid) throws RequestException {
+    /**
+     * Looks up for a batch of accounts.
+     *
+     * @param batch An array containing all accounts ids.
```
"all account ids" (singular, not plural)

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -108,27 +218,79 @@ public Account lookupAccount(UUID uuid) throws RequestException {
     }
 
+    /**
+     * Looks up for a batch of accounts.
```
No "for"

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -108,27 +218,79 @@ public Account lookupAccount(UUID uuid) throws RequestException {
+    /**
+     * Looks up for a batch of accounts.
+     *
+     * @param batch An {@link com.tigerbeetle.UUIDsBatch} containing all accounts ids.
```
"all account ids"

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -108,27 +218,79 @@ public Account lookupAccount(UUID uuid) throws RequestException {
     }
 
+    /**
+     * Looks up for a batch of accounts asynchronously.
```
No "for"

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -108,27 +218,79 @@ public Account lookupAccount(UUID uuid) throws RequestException {
+     * Looks up for a batch of accounts asynchronously.
+     *
+     * @see Client#lookupAccounts(UUID[])
+     * @param batch An array containing all accounts ids.
```
"all account ids"

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Client.java:
```diff
@@ -141,30 +303,84 @@ public CreateTransferResult createTransfer(Transfer transfer) throws RequestExce
     }
 
+    /**
+     * Looks up for a single transfer.
```
no "for"

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Transfer.java:
```diff
@@ -54,43 +54,102 @@ public Transfer() {
+     *
+     * @return An {@link java.util.UUID} representing an integer-encoded UUIDv4 or any other unique
+     *         128-bit integer.
+     */
```
Same comment as for account. These aren't required to be UUIDs

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Transfer.java:
```diff
@@ -54,43 +54,102 @@ public Transfer() {
 
+    /**
+     *
+     * <p>
```
Missing the first part of the doc? Or did I not add more detail in my PR

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Transfer.java:
```diff
@@ -54,43 +54,102 @@ public Transfer() {
+     * @return An {@link java.util.UUID} representing an integer-encoded UUIDv4 or any other unique
+     *         128-bit integer.
+     */
     public UUID getUserData() {
```
Same comment about not required to be a uuid

eatonphil commented on src/tigerbeetle-java/src/main/java/com/tigerbeetle/Transfer.java:
```diff
@@ -111,60 +184,131 @@ public void setPendingId(UUID pendingId) {
+     * Gets the behavior during transfers.
+     *
+     * @see com.tigerbeetle.TransferFlags
+     * @return A 16-bits unsigned integer bit mask.
```
Can we turn these into meaningful Java enums or something in the client? Rather than requiring the user to do the work

batiati replied:
We had that but it was removed here: https://github.com/tigerbeetledb/tigerbeetle-java/commit/7ba7bbb0c6ced42bcf4a7d9f408041b884d16cc8#diff-49dd611e9452f59780634d2d13c834c281bf532a482553dab906dd928922f282

- Removed non-ergonomic enums AccountFlags and TransactionFlags in favor of a simple integer constant.
Maybe implementing a helper to combine the flags would be a better solution

Setting the flag is pretty straightforward since Java supports bitwise OR with `A | B`, but reading individual flags could use a helper.

batiati committed:
Javadoc corrections

batiati committed:
Adds test coverage thresholds

- GHA will fail with less than 95% of tested intructions or less than 90% of tested branches.
We have low branch ratio due to asserts on unexpected values that are unreachable.

- The current values on this commit are 96% and 93% respectively.
